app/src/main/java/com/example/countryinfo/CountryDetailsAdapter.kt

```diff
@@ -0,0 +1,162 @@
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
+        val view = LayoutInflater.from(parent.context)
+            .inflate(R.layout.country_details_card, parent, false)
+        return ViewHolder(view, countryDetails, listener)
+    }
+
+    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
+        holder.bind(position)
+    }
+
+    override fun getItemCount(): Int {
+        return DetailsStates.values().size
+    }
+
+    class ViewHolder(
+        view: View,
+        private val countryDetails: CountryDetails,
+        private val listener: OnItemClickListener
+    ) : RecyclerView.ViewHolder(view), View.OnClickListener {
+
+        var viewState: DetailsStates? = null
+        var dialogMessage = emptyList<String>()
+
+        init {
+            itemView.setOnClickListener(this)
+        }
+
+        override fun onClick(v: View?) {
+            viewState?.let { listener.onItemClick(it, dialogMessage) }
+        }
+
+        interface OnItemClickListener {
+            fun onItemClick(state: DetailsStates, dialogMessage: List<String>)
+        }
+
+        fun bind(pos: Int) {
+            if (pos % 2 == 0) {
+                itemView.circleImage.setImageResource(R.drawable.circle1)
+            } else {
+                itemView.circleImage.setImageResource(R.drawable.circle2)
+            }
+            when (pos) {
+                0 -> {
+                    itemView.labelText.text = DetailsStates.COUNTRY_NAME.title
+                    itemView.titleText.text = countryDetails.name
+                }
+                1 -> {
+                    itemView.labelText.text = DetailsStates.CAPITAL.title
+                    itemView.titleText.text = countryDetails.capital
+                }
+                2 -> {
+                    itemView.labelText.text = DetailsStates.REGION.title
+                    itemView.titleText.text = countryDetails.region
+                }
+                3 -> {
+                    itemView.labelText.text = DetailsStates.POPULATION.title
+                    itemView.titleText.text = formatPopulation(countryDetails.population)
+                }
+                4 -> {
+                    itemView.labelText.text = DetailsStates.CURRENCIES.title
+                    bindMultiple(DetailsStates.CURRENCIES, countryDetails)
+                }
+                5 -> {
+                    itemView.labelText.text = DetailsStates.LANGUAGES.title
+                    bindMultiple(DetailsStates.LANGUAGES, countryDetails)
+                }
+                6 -> {
+                    itemView.labelText.text = DetailsStates.TIMEZONES.title
+                    bindMultiple(DetailsStates.TIMEZONES, countryDetails)
+                }
+                7 -> {
+                    itemView.labelText.text = DetailsStates.CALLING_CODES.title
+                    bindMultiple(DetailsStates.CALLING_CODES, countryDetails)
+                }
+            }
+        }
```

Comment:
While this is a working approach adapter contains a lot of view state mapping logic which should be done in the ViewModel layer. Another approach could be is to create DetailsViewHolderState class like this:
```
data class DetailsViewHolderState(title: String, info: List<String>, isInfoHighlighted: Boolean)
```

Adapter will maintain a collection of `DetailsViewHolderState` objects. The mapping logic will be in the ViewModel layer in this way.

Reply:
All done.